grpc/server_test.go

```diff
@@ -0,0 +1,293 @@
+			if err != nil {
+				t.Fatalf("net.Listen error: %s", err)
+			}
+			defer lis.Close()
+
+			opt := grpc.UnaryInterceptor(sentrygrpc.UnaryServerInterceptor(tt.opts))
+			server := grpc.NewServer(opt)
+			defer server.Stop()
+
+			grpchealth.RegisterHealthServer(
+				server,
+				&mockHealthServer{handler: tt.handler},
+			)
+
+			go func() {
+				if err := server.Serve(lis); err != nil {
+					t.Errorf("grpc serve error: %s", err)
+				}
+			}()
+
+			conn, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure())
+			if err != nil {
+				t.Fatalf("grpc.Dial error: %s", err)
+			}
+			defer conn.Close()
+
+			client := grpchealth.NewHealthClient(conn)
+
+			req := &grpchealth.HealthCheckRequest{}
+			_, err = client.Check(tt.ctx, req)
+			if w, g := tt.code, status.Code(err); w != g {
+				t.Fatalf("status mismatch: want %s, got %s", w, g)
+			}
+
+			if !sentry.Flush(time.Second) {
+				t.Fatal("sentry.Flush timed out")
+			}
+
+			opts := cmp.Options{
+				cmpopts.IgnoreFields(
+					sentry.Event{},
+					"Contexts",
+					"EventID",
+					"Extra",
+					"Platform",
+					"Sdk",
+					"ServerName",
+					"Tags",
+					"Timestamp",
+				),
+				cmpopts.IgnoreFields(
+					sentry.Request{},
+					"Env",
+				),
+				cmpopts.IgnoreFields(
+					sentry.Exception{},
+					"Stacktrace",
+				),
+			}
+
+			if d := cmp.Diff(tt.events, events, opts); d != "" {
+				t.Fatalf("events mismatch (-want +got):\n%s", d)
+			}
+		})
+	}
+}
+
+type mockHealthServer struct {
+	grpchealth.UnimplementedHealthServer
+
+	handler func(
+		context.Context,
+		*grpchealth.HealthCheckRequest,
+	) (*grpchealth.HealthCheckResponse, error)
+}
+
+func (m *mockHealthServer) Check(
+	ctx context.Context,
+	req *grpchealth.HealthCheckRequest,
+) (*grpchealth.HealthCheckResponse, error) {
+	return m.handler(ctx, req)
+}
+
+func TestReportOnCodes(t *testing.T) {
+	for _, tt := range []struct {
+		name  string
+		err   error
+		codes []codes.Code
+		want  bool
+	}{
+		{
+			name:  "returns true on code match",
+			err:   status.Error(codes.Aborted, ""),
+			codes: []codes.Code{codes.Aborted},
+			want:  true,
+		},
+		{
+			name:  "returns false on code mismatch",
+			err:   status.Error(codes.Aborted, ""),
+			codes: []codes.Code{codes.Canceled},
+			want:  false,
+		},
+	} {
+		tt := tt
+
+		t.Run(tt.name, func(t *testing.T) {
+			if w, g := tt.want, sentrygrpc.ReportOnCodes(tt.codes...)(tt.err); w != g {
+				t.Fatalf("ReportOnCodes: want %t, got %t", w, g)
+			}
+		})
+	}
+}
+
+func ExampleUnaryServerInterceptor() {
+	opts := sentrygrpc.UnaryServerInterceptorOptions{
+		// Reports on OutOfRange or Internal error.
+		ReportOn: sentrygrpc.ReportOnCodes(
+			codes.OutOfRange,
+			codes.Internal,
+		),
+		// Recovers from panic, reports it and returns internal error.
+		Repanic: false,
+	}
+
+	// This middleware sets *sentry.Hub to context. You can set user to
+	// hub's scope in the later interceptor for example.
+	sentry := sentrygrpc.UnaryServerInterceptor(opts)
```

Comment:
@shouichi do you think the default options is what will be most used?

I know `sentryhttp` and friends follow this pattern of taking an option struct, but for new code I often consider if we can use the [functional options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

As an example, there's a similar interceptor in [opentracing-contrib/go-grpc](https://github.com/opentracing-contrib/go-grpc/blob/4a6760c71486b5e9aa6faf22b11d579e13a119d9/server.go#L27).

```go
func OpenTracingServerInterceptor(tracer opentracing.Tracer, optFuncs ...Option) grpc.UnaryServerInterceptor {
```

What the different style would allow is for:

- No empty struct as required argument
  ```go
  sentrygrpc.UnaryServerInterceptor()
  ```
  vs
  ```go
  sentrygrpc.UnaryServerInterceptor(sentrygrpc.UnaryServerInterceptorOptions{})
  ```

The `grpc` library itself uses (a form of) functional options.

- Possibility to extend the options and having defaults that do not necessarily match Go zero values.

Reply:
> do you think the default options is what will be most used?

Currently yes, but in the future, I'm not sure. The functional options pattern might be friendlier to users. My personal preference is to be consistent. So if the future direction is to use the functional options pattern, I'm happy to use it.

I often find myself wondering about options struct vs functional options. It seems that the go stdlib embraces zero values but I see a lot of usage of functional options pattern in the wild. It's hard to pick one!

Reply:
> It seems that the go stdlib embraces zero values but I see a lot of usage of functional options pattern in the wild.

The pattern is something that emerged many years later when the stdlib was already solidified by the Go 1 compatibility promise.
But new experiments like the [new HTTP client](https://github.com/bradfitz/exp-httpclient) use functional options.

Reply:
Cool! Didn't know that.

OK, I'll change to use the functional options pattern.